### ISSUE TITLE
Connection event doesn't always fire

### ISSUE THREAD

**Issue by goldfire** [primus/primus#391]
We've run into a very strange issue that is even more complicated due to the fact that we can't reproduce it. However, we've received significant numbers of reports of the issue and have been able to gather quite a bit of data and logs from the instances it has happened in production.

For around 10-20% of our visitors, the `connection` event never fires on the server when first loading the page. The problem happens to these same people consistently, and never to anyone else. It doesn't seem to matter what browser or OS they are using. The only way they can get the connection to go through is by clicking in the address bar and hitting enter, rather than doing a normal refresh of the page. In addition, they actually are connected, because they successfully send/receive pings/pongs and receive data broadcast from other clients.

We've been debugging this all week and have run out of ideas. Any thoughts on what could possibly cause this or where else to look?

Edit: We are using the `ws` module through primus.

**Comment by lpinca** [comment]
I honestly have no idea, this is kinda strange. What strikes me as odd is the fact that the client connection works.

Do `diagnostics` logs give any hints on what happens?
```sh
DEBUG=primus:spark,primus:transformer node app.js
```
Did you try to use another transformer? For example does the same happen if you try to use `faye`?

**Comment by goldfire** [comment]
I hadn't tried the diagnostics logs yet, switching that on in production now.

**Comment by goldfire** [comment]
I had originally diagnosed it incorrectly due to the large volume of logs. After cleaning up our logs, a completely different picture was painted and we were able to get it sorted. I don't believe there's any issue with Primus (possibly with ws, but we worked around it).

**Comment by 3rd-Eden** [comment]
@goldfire Thanks for the update? What was the actual issue if I may ask?

**Comment by goldfire** [comment]
@3rd-Eden Right, sorry, was waiting to make sure our update actually fixed the issue (since we were still never able to reproduce on our end).

When we receive the connection event, we immediately send a message back to that client with some data to get their client fully setup. The readyState was set to OPEN at this point, and we tracked it all the way to the `ws` send method in `Sender.js` where everything seemed to log correctly. However, the client never received that message (but they would receive other broadcasts after that one).

We were running out of ideas, so I figured I'd add a 100ms timeout around that initial broadcast just to see what would happen, and so far all of the players reporting the issue have said they can't get it to happen anymore. Any ideas why that would be?
